### PR TITLE
Fix more segfaults on NetPlay quit

### PR DESCRIPTION
**Pull request by Techjar**
Basically everything here was race conditions in Qt callbacks, so I changed the client/server instances to `std::shared_ptr` and added null checks. It checks that the object exists in the callback, and the `shared_ptr` ensures it doesn't get destroyed until we're done with it.

MD5 check would also cause a segfault if you quit without cancelling it first, which was pretty silly.

### CODE REVIEW

File: Source/Core/Core/NetPlayClient.cpp [hunk 1]
```diff
@@ -57,6 +57,10 @@ NetPlayClient::~NetPlayClient()
 
   if (m_is_connected)
   {
+    m_should_compute_MD5 = false;
+    m_dialog->AbortMD5();
+    if (m_MD5_thread.joinable())
```
Comment (JonnyH) [inline Source/Core/Core/NetPlayClient.cpp]: Doesn't .detach() set joinable to false, so this will never be joined?

Comment (Techjar) [reply Source/Core/Core/NetPlayClient.cpp]: Yeah you're right, no point in joining the thread. It's also now randomly segfaulting in another place, so needs more null checks.

File: Source/Core/DolphinQt/MainWindow.cpp [hunk 2]
```diff
@@ -1088,7 +1088,7 @@ bool MainWindow::NetPlayJoin()
   const std::string nickname = Config::Get(Config::NETPLAY_NICKNAME);
 
   // Create Client
-  const bool is_hosting_netplay = Settings::Instance().GetNetPlayServer() != nullptr;
+  const bool is_hosting_netplay = !!Settings::Instance().GetNetPlayServer();
```
[the remark this answers is not in the document]
Comment (Techjar) [reply Source/Core/DolphinQt/MainWindow.cpp]: Ah, I didn't realize `shared_ptr` has an `operator!=` overload.

File: Source/Core/Core/NetPlayClient.cpp (hunk 1 above)
Comment (Techjar) [reply Source/Core/Core/NetPlayClient.cpp]: This has been addressed by handling the thread more sanely.
